### PR TITLE
KAFKA-8391: Improved the Connect integration tests to make them less flaky

### PR DESCRIPTION
**Pull request by rhauch**
Added the ability for the connector handles and task handles, which are used by the monitorable source and sink connectors used to verify the functionality of the Connect framework, to record the number of times the connector and tasks have each been started, and to allow a test to obtain a `RestartLatch` that can be used to block until the connectors and/or tasks have been restarted a specified number of types.

Typically, a test will get the `ConnectorHandle` for a connector, and call the `ConnectorHandle.expectedRestarts(int)` method with the expected number of times that the connector and/or tasks will be restarted, and will hold onto the resulting `RestartLatch`. The test will then change the connector (or otherwise cause the connector to restart) one or more times as desired, and then call `RestartLatch.await(long, TimeUnit)` to block the test up to a specified duration for the connector and all tasks to be started the specified number of times.

This commit also increases several of the maximum wait times used in other integration tests. It doesn’t hurt to potentially wait longer, since most test runs will not need to wait the maximum amount of time anyway. However, in the rare cases that do need that extra time, waiting a bit more is fine if we can reduce the flakiness and minimize test failures that happened to time out too early.

Unit tests were added for the new `RestartLatch` and `StopAndStartCounter` utility classes. This PR only affects the tests and does not affect any runtime code or API.

**This should be merged on `trunk` and backported to the `2.3.x` branch.**

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java
Comment (rhauch) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java:178]: Note that these did not add any value: the results were never verified in this test. But more importantly, because this test is restarting the connector and tasks, it's really difficult to measure the number of commits that will happen after the connector is restarted. Therefore, it makes no sense to use these methods, since we're using a consumer to verify the number of records that the connector writes to the alternative topic upon connector/task restart.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java
Comment (abbccdda) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java]: could we add a unit like `timeoutMs`?

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/StartAndStopCounterTest.java
Comment (abbccdda) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/StartAndStopCounterTest.java]: nit: new line

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java
Comment (abbccdda) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java]: nit: spaces

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java
Comment (abbccdda) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java]: nit: format

**Review by abbccdda — COMMENTED** [review]
Thank you for fixing this! Just some minor style comments

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java]: I was trying to avoid lots of changes, but I agree it's useful. We do want to keep these test APIs backward compatible because the integration test framework is used outside of AK, so I'll add overloaded methods that take the unit.

**Comment by rhauch** [comment]
Thanks for the review, @abbccdda. I've fixed the PR to include your suggested fixes, including adding a missing unit test. I would appreciate another review.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java
Comment (wicknicks) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java]: do we need both the `expectedStarts` and `expectedRestarts`? It seems like the former should be just one more than the other.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java
Comment (wicknicks) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java]: by doing this, we are assuming that a connector restart implies a task restart. that might not be always true. adding workers might restart some tasks but not others (esp., with the new rebalancing algorithm). do you think we should allow users to be able to increment the expected number of restarts for individual tasks. if so, does this dependency still hold?

**Review by wicknicks — COMMENTED** [review]
Excellent PR, @rhauch. Couple of questions.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java
Comment (abbccdda) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java]: Seems L64-67 have been reused multiple times. We could either refactor a setup function or just make the wait parameters `100, 10, 200` as static parameters if we need to tune them later.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java
Comment (abbccdda) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java]: new line

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java
Comment (abbccdda) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java]: Sounds interesting, cc @kkonstantine

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/StartAndStopCounterTest.java
Comment (abbccdda) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/StartAndStopCounterTest.java]: nit: remove commented logic

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/StartAndStopCounterTest.java
Comment (abbccdda) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/StartAndStopCounterTest.java]: nit: new line

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java]: nit: `this.` is missing and although not strictly needed probably better to be used

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java]: nit: `this.` is missing and although not strictly needed probably better to be used

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java]: not required in junit 4 and later. `waiters` get GCed

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java
Comment (kkonstantine) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java]: It's good for small test cases like these to avoid the extra redirection to a helper method and keep them concise and self-descriptive as they are now.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java
Comment (kkonstantine) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java]: Actually I have the opposite comment to make. 
I wonder how useful it is to make this very generic and add a lot of boilerplate. In practice we'll never use in tests anything other than milliseconds. To the contrary we should discourage anyone doing so, because this is a real-time (not mocked) delay and delays in the order of multiple seconds are problematic. Even if we needed a few minutes (which I hope we'll never will), `TimeUnit.MILLISECONDS` is more than enough.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java:187]: Same comment as above, to avoid excessive parameterization.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java
Comment (kkonstantine) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java]: I'd assume this is meant to be used to assert reconfiguration. But I agree that it can be abused or be misleading. Also is it correct to assume that `expectedRestarts` means at least that many?

**Review by kkonstantine — COMMENTED** [review]
Thanks @rhauch 
I hope this help making some of the integrations tests more robust
I skimmed through the code and left a few comments. 

Another concern I have is that the implementation seems to add support for dependent latches but I don't see this functionality being used. I wonder if the increased complexity this concept introduces will ever be useful.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java]: I'll change this to `expectedStarts` and `expectedStops`, and remove `expectedRestarts`. But note that all of these start counting when a test calls the method. So tests basically do the following after the state gets to some interesting point:
1) call one of these methods with the (minimum) number of starts (or stops) that should be expected
2) do anything that causes the connector (and/or tasks) to be stopped and/or started
3) call the latch to await until the connector (and optionally tasks) are started and/or stopped the expected number of times, or fail if that doesn't happen within a specific timeout.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java]: Yes, I've updated the JavaDoc to denote this is a minimum number, and as mentioned above I've removed the `expectedRestarts` methods and now we have `expectedStarts` and `expectedStops` (if you really care about stops).

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java]: Agreed. I've gotten dinged in previous PRs about using `this.` when it is not required, but I prefer the consistency here, so I'm adding it as you suggest.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java]: Agreed.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java]: I agree. I find it harder to follow helper method in situations like this where some of the calls or order of calls do vary amongst the tests.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java]: fixed

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java]: I guess I disagree. I don't think we need minutes, but we may need 10s of seconds, and it's too easy to mistype or misread the number of zeros in `10000`. I would much rather use `awaitRecords(10, TimeUnit.SECONDS)`. And this is the convention used in all of the JDK for awaiting for something to happen, and I think sticking with JDK conventions is good.

**Review by rhauch — COMMENTED** [review]
I've updated the PR and addressed most of the comments. I do disagree about whether to include methods with `TimeUnit` parameters, as noted below.

**Comment by rhauch** [comment]
The first build failure doesn't appear to be related to this PR (it's a Gradle error), but the second has a failure of [another flaky integration test](https://issues.apache.org/jira/browse/KAFKA-8661?jql=project%20%3D%20KAFKA%20AND%20text%20~%20%22RebalanceSourceConnectorsIntegrationTest%22). Will look into this more. Would appreciate another review of these changes so far.

**Comment by ijuma** [comment]
Will this fix RebalanceSourceConnectorsIntegrationTest.testReconfigConnector and RebalanceSourceConnectorsIntegrationTest#testStartTwoConnectors too? `testReconfigConnector`, in particular, fails a lot.

@rhauch are you actively working on this? If not, I think we have to disable the 3 flaky tests until this lands. These tests have been flaky for too long now and it slows down everyone working in Apache Kafka.

@mjsax Would you be able to help review this PR, seems like @rhauch was looking for a committer to review it.

**Comment by mjsax** [comment]
I am not familiar with this part of the code, but I can give it a try. If @kkonstantine and @wicknicks approve the PR, I would feel confident to merge it.

@rhauch I might be good to rebase this PR to improve build stability -- we merged a fix for a flaky test last night

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java
Comment (kkonstantine) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatch.java]: I agree, conflicting comments are a reality, hopefully a rare one. I believe I suggest using `this` only within constructors.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java
Comment (kkonstantine) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/RestartLatchTest.java]: This comment was resolved without a follow up or a change. I know setting to `null` is not required here, but if it's a matter of style or you prefer using this anyway I don't want to insist. But if I've missed something it'd be helpful to know.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java
Comment (kkonstantine) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java]: I don't feel too strong about this change. My points where a) to discourage using seconds for delays b) that this is not a general library such as the JDK or other in which I'd be all up for the fully customizable method version. Seems fine to leave as is though and not block this PR on this suggestion.

**Review by kkonstantine — APPROVED** [review]
Thanks for addressing my comments @rhauch ! 
This LGTM now and it's definitely an improvement, even if flaky test might need some more work. 

I see the dependent latches stay, I guess we'll see in practice how useful this construct will be. 

@mjsax I'd suggest merging this PR, ideally backporting it to 2.3.x too as you also mentioned.

**Comment by rhauch** [comment]
@kkonstantine, thanks for the review. 

@mjsax, I just rebased this PR on latest `trunk`, and building and running all of the Connect integration tests locally passed for me several times in a row. Let's see what this does.

**Comment by gwenshap** [comment]
LGTM. Merged to master. I couldn't cherrypick cleanly to 2.3.

**Pull request closed without merging**
